rl_bakery/agent_trainer/agents.py

```diff
@@ -0,0 +1,173 @@
+import abc
+import tensorflow as tf
+from tf_agents.utils import common
+from tf_agents.networks import q_network
+from tf_agents.agents.dqn import dqn_agent
+from tf_agents.agents import ddpg
+from tf_agents.trajectories import time_step as ts
+import logging
+
+from rl_bakery.agent_trainer.config import Optimizer
+
+logger = logging.getLogger(__name__)
+
+
+class AgentTrainer(metaclass=abc.ABCMeta):
+    """
+    This set ups an Agent to be trained.
+    """
+
+    def __init__(self, data_spec, config):
+        self._data_spec = data_spec
+        self._config = config
+
+    @abc.abstractmethod
+    def init_agent(self):
+        pass
+
+
+class TFAgentTrainer(AgentTrainer):
```

Comment:
I don't think inheritance is the right way to go here. Using it led to the creation of a long hierarchy of classes 
Level 1: TFAgentTrainer: Builds optimizer
Level 2: QAgent: Builds network 
Level 3: DQNAgent: Builds a DQN

Disadvantages of inheritance here:
* Coupling between the classes. If any code changes in level 1 or 2, all classes at level 3 will be impacted. 
* The behaviour inherited is set at compile-time. -> it would not be possible to change the parent class at runtime.  (e.g. the optimizer setup cannot change at runtime)

I think we should just have specific "AgentInitializers" be direct implementations of "AgentInitializersInterface". We could create two of these implementations: TFAgentDQNInitializer, and TFAgentDDPGInitializer for example.